File: src/main/java/com/chaco/algorithms/dp/Distance.java
package com.chaco.chao.algorithms.dp;

import java.util.Objects;

/**
 * @author zhaopeiyan
 * @date 2021/12/9 5:28 下午
 */
public class Distance {

    /**
     * 给定两个字符串
     * str1=horse
     * str2=ros
     * 计算出变为相等修改的最短路径
     *
     * @param word1
     * @param word2
     * @return
     */
    public static int editDistance(String word1, String word2) {
        int[][] dp = new int[word1.length() + 1][word2.length() + 1];
        // 第一行
        for (int j = 1; j <= word2.length(); j++) {
            dp[0][j] = dp[0][j - 1] + 1;
        }
        // 第一列
        for (int i = 1; i <= word1.length(); i++) {
            dp[i][0] = dp[i - 1][0] + 1;
        }

        for (int i = 1; i <= word1.length(); i++) {
            for (int j = 1; j <= word2.length(); j++) {
                if (Objects.equals(i, -1)) {
                    return i + 1;
                }
                if (Objects.equals(j, -1)) {
                    return j + 1;
                }
                if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = Math.min(Math.min(dp[i - 1][j], dp[i][j - 1]), dp[i - 1][j - 1]) + 1;
                }
            }
        }
        return dp[word1.length()][word2.length()];
    }

    public static void main(String[] args) {
        int i = Distance.editDistance("horse", "ros");
        int i1 = Distance.editDistance("apple", "rad");
        int i2 = Distance.editDistance("intention", "execution");
        System.out.println(i);
    }
}

File: src/main/java/com/chaco/algorithms/dp/Distance.go
package dp

import "math"

func minDistance(word1 string, word2 string) int {
	var dp = make([][]int, len(word1)+1, len(word2)+1)
	for i := 1; i < len(word2); i++ {
		dp[0][i] = i
	}
	for j := 1; j < len(word1); j++ {
		dp[j][0] = j
	}
	for i := 1; i <= len(word1); i++ {
		for j := 1; j <= len(word2); j++ {
			if i == -1 {
				return i + 1
			}
			if j == -1 {
				return j + 1
			}
			if word1[i-1] == word2[j-1] {
				dp[i][j] = dp[i-1][j-1]
			} else {
				dp[i][j] = int(math.Min(float64(dp[i-1][j-1]), math.Min(float64(dp[i][j-1]), float64(dp[i-1][j])))) + 1
			}
		}
	}
	return dp[len(word1)][len(word2)]
}

func Min(args ...int) int {
	min := args[0]
	for _, item := range args {
		if item < min {
			min = item
		}
	}
	return min
}

func Max(args ...int) int {
	max := args[0]
	for _, item := range args {
		if item > max {
			max = item
		}
	}
	return max
}

func minDistance1(word1 string, word2 string) int {
	n1, n2 := len(word1), len(word2)
	dp := make([][]int, n1+1)
	for i := range dp {
		dp[i] = make([]int, n2+1)
	}
	for i := 1; i < len(word2); i++ {
		dp[0][i] = dp[0][i-1] + 1
	}
	for j := 1; j < len(word1); j++ {
		dp[j][0] = dp[j-1][0] + 1
	}
	for i := 1; i <= len(word1); i++ {
		for j := 1; j <= len(word2); j++ {
			if word1[i-1] == word2[j-1] {
				dp[i][j] = dp[i-1][j-1]
			} else {
				dp[i][j] = Min(dp[i-1][j-1], Min(dp[i][j-1], dp[i-1][j])) + 1
			}
		}
	}
	return dp[n1][n2]
}
